File: android/app/src/main/java/com/medwand/poc/bridge/Protocol.kt
package com.medwand.poc.bridge

import com.medwand.poc.device.ActiveSensor
import com.medwand.poc.device.Capture
import com.medwand.poc.device.CameraFrame
import com.medwand.poc.device.DeviceSnapshot
import com.medwand.sdk_core.Core.MedWandReading
import org.json.JSONObject

/**
 * Serialization for the bridge envelope (bridge doc §2). All web ↔ native JSON
 * passes through here; nothing else builds protocol strings by hand.
 */
object Protocol {

    /** Every native → web message is a uniform `{ event, data }` envelope. */
    fun event(event: String, data: JSONObject): String =
        JSONObject().apply {
            put("event", event)
            put("data", data)
        }.toString()

    /** A command reply is just a `reply` event whose data echoes the command id. */
    fun replyOk(id: String, result: JSONObject = JSONObject()): String =
        event("reply", JSONObject().put("id", id).put("ok", true).put("result", result))

    fun replyError(id: String, code: String, message: String): String =
        event(
            "reply",
            JSONObject().put("id", id).put("ok", false)
                .put("error", JSONObject().put("code", code).put("message", message)),
        )

    fun errorEvent(code: String, message: String): String =
        event("error", JSONObject().put("code", code).put("message", message))

    fun snapshotJson(snapshot: DeviceSnapshot): JSONObject = JSONObject().apply {
        put("device", snapshot.device)
        put("reading", snapshot.reading)
        put("activeSensor", snapshot.activeSensor?.let { activeSensorJson(it) } ?: JSONObject.NULL)
        put(
            "capabilities",
            JSONObject().apply {
                put("thermometer", snapshot.capabilities.thermometer)
                put("pulseOximeter", snapshot.capabilities.pulseOximeter)
                put("ecg", snapshot.capabilities.ecg)
                put("stethoscope", snapshot.capabilities.stethoscope)
                put("camera", snapshot.capabilities.camera)
            },
        )
        put(
            "deviceInfo",
            JSONObject().apply {
                put("udi", snapshot.deviceInfo.udi ?: JSONObject.NULL)
                put("firmwareVersion", snapshot.deviceInfo.firmwareVersion ?: JSONObject.NULL)
                put("generation", snapshot.deviceInfo.generation ?: JSONObject.NULL)
                put("comPort", snapshot.deviceInfo.comPort ?: JSONObject.NULL)
                put("vendorId", snapshot.deviceInfo.vendorId ?: JSONObject.NULL)
                put("productId", snapshot.deviceInfo.productId ?: JSONObject.NULL)
            },
        )
    }

    /**
     * The active sensor as `{ sensor, ...settings }`. `sensor` is the
     * discriminator the frontend narrows on; the camera and stethoscope add their
     * live control fields, plain sensors add nothing.
     */
    private fun activeSensorJson(active: ActiveSensor): JSONObject = JSONObject().apply {
        put("sensor", active.sensor)
        when (active) {
            is ActiveSensor.Camera -> {
                put("mode", active.mode)
                put("monitoring", active.monitoring)
                put("ledMax", active.ledMax)
                put("ledIntensity", active.ledIntensity)
                put("ledAdjustable", active.ledAdjustable)
                put("autoFocusAvailable", active.autoFocusAvailable)
                put("manualFocusAvailable", active.manualFocusAvailable)
                put("manualFocusEnabled", active.manualFocusEnabled)
                put("focusValue", active.focusValue)
                put("focusValueMax", active.focusValueMax)
            }
            is ActiveSensor.Stethoscope -> put("mode", active.mode)
            is ActiveSensor.Plain -> {}
        }
    }

    /**
     * A `reading` event: `{ sensor, reading }` where `sensor` is our sensor name
     * and `reading` is the SDK's `MedWandReading` (§5).
     */
    fun readingEventJson(sensor: String, reading: MedWandReading): JSONObject = JSONObject().apply {
        put("sensor", sensor)
        put("reading", readingJson(reading))
    }

    /**
     * The SDK's `MedWandReading` serialized **verbatim** — every field, including
     * empty strings and nulls — so the frontend (and debugging) see exactly what
     * the device emitted: `index`/`count` position, raw `ecgData`, and the rest.
     * Scalars stay as the SDK's own strings; the frontend parses them.
     */
    fun readingJson(reading: MedWandReading): JSONObject = JSONObject().apply {
        put("sensorType", reading.sensorType ?: JSONObject.NULL)
        put("timeStamp", reading.timeStamp?.toString() ?: JSONObject.NULL)
        put("status", reading.status ?: JSONObject.NULL)
        put("index", reading.index)
        put("count", reading.count)
        put("tempAmbient", reading.tempAmbient ?: JSONObject.NULL)
        put("tempObject", reading.tempObject ?: JSONObject.NULL)
        put("pulseRate", reading.pulseRate ?: JSONObject.NULL)
        put("spo2", reading.spo2 ?: JSONObject.NULL)
        put("ecgData", reading.ecgData ?: JSONObject.NULL)
    }

    /**
     * A recorded capture as a `capture` event. The SDK already produces a
     * base64 `data:` URI, so it is forwarded as-is for the frontend to use
     * directly as an <img>/<audio> source.
     */
    fun captureJson(capture: Capture): JSONObject = JSONObject().apply {
        put("sensor", capture.sensor)
        capture.mode?.let { put("mode", it) }
        put("dataUri", capture.dataUri)
    }

    /**
     * A live camera preview frame as a `frame` event: a JPEG `data:` URI plus the
     * frame dimensions. This is the **base64 fallback** used only when the WebView
     * lacks array-buffer support; the fast path ships the JPEG bytes raw over the
     * binary channel instead. High-frequency, so kept off the `reading`/`capture`
     * channels either way.
     */
    fun cameraFrameJson(frame: CameraFrame, dataUri: String): JSONObject = JSONObject().apply {
        put("sensor", "camera")
        put("mode", frame.mode)
        put("dataUri", dataUri)
        put("width", frame.width)
        put("height", frame.height)
    }
}

File: frontend/src/bridge/protocol.ts
// Wire protocol shared with the native bridge (see docs/medwand-bridge.md).

export type SensorId = "thermometer" | "pulseOximeter" | "ecg" | "stethoscope" | "camera";
export type StethMode = "heart" | "lungs" | "bowel";
export type CameraMode = "dermatoscope" | "otoscope";

// Per-sensor start options (mirror of the native SensorOptions). Only the
// stethoscope and camera need one; the relevant field is picked by sensor.
export interface SensorOptions {
  stethMode?: StethMode;
  cameraMode?: CameraMode;
}

export type ErrorCode =
  | "SENSOR_BUSY"
  | "NOT_CONNECTED"
  | "USB_PERMISSION_DENIED"
  | "LICENSE_INVALID"
  | "DEVICE_ERROR";

export interface BridgeError {
  code: ErrorCode | string;
  message: string;
}

export interface DeviceState {
  device: string; // detached | attached | connecting | connected
  reading: string; // idle | active | recording
  // The running sensor plus its live, user-settable state (null when idle).
  activeSensor: ActiveSensor | null;
  capabilities: {
    thermometer: boolean;
    pulseOximeter: boolean;
    ecg: boolean;
    stethoscope: boolean;
    camera: boolean;
  };
  deviceInfo: {
    udi: string | null;
    firmwareVersion: string | null;
    generation: string | null;
    comPort: string | null;
    vendorId: string | null;
    productId: string | null;
  };
}

// The active sensor is a tagged union discriminated on `sensor`: the camera and
// stethoscope carry their live control state, the rest carry nothing.
export type ActiveSensor = CameraSensorState | StethoscopeSensorState | PlainSensorState;

// Live camera control state (present only while the camera is active). Ranges
// (`ledMax`, `focusValueMax`, availability flags) sit here with the current
// values because they're session-scoped and can differ by mode.
export interface CameraSensorState {
  sensor: "camera";
  mode: "off" | CameraMode; // active preview mode
  monitoring: boolean;
  ledMax: number;
  ledIntensity: number;
  ledAdjustable: boolean;
  autoFocusAvailable: boolean;
  manualFocusAvailable: boolean;
  manualFocusEnabled: boolean;
  focusValue: number;
  focusValueMax: number;
}

export interface StethoscopeSensorState {
  sensor: "stethoscope";
  mode: StethMode;
}

export interface PlainSensorState {
  sensor: "thermometer" | "pulseOximeter" | "ecg";
}

// The SDK's MedWandReading, forwarded verbatim (all fields, incl. nulls).
// Scalars are strings; the frontend parses them. ECG's `ecgData` is the raw
// whitespace-separated dimensionless waveform.
export interface SdkReading {
  sensorType: string | null;
  timeStamp: string | null;
  status: string | null;
  index: number;
  count: number;
  tempAmbient: string | null;
  tempObject: string | null;
  pulseRate: string | null;
  spo2: string | null;
  ecgData: string | null;
}

// The `reading` event: our sensor name + the raw SDK reading.
export interface ReadingEvent {
  sensor: SensorId;
  reading: SdkReading;
}

// A recorded artifact, forwarded from the SDK as a base64 `data:` URI that can
// be used directly as an <img>/<audio> source.
export interface Capture {
  sensor: SensorId | string;
  mode?: string;
  dataUri: string;
}

// A live camera preview frame as it arrives over the JSON string channel: a JPEG
// `data:` URI plus its dimensions. Used by the browser mock and by the native
// base64 fallback. The native fast path instead delivers the JPEG bytes as a raw
// binary `ArrayBuffer` message (no envelope) — see MedWandBridge.
export interface WireFrameEvent {
  sensor: "camera";
  mode: string;
  dataUri: string;
  width: number;
  height: number;
}

// The normalized frame handed to listeners, regardless of transport: a
// ready-to-use image `src` (a base64 `data:` URI on the string path, an object
// URL on the binary path).
export interface FrameEvent {
  sensor: "camera";
  src: string;
}

// Reply payload, carried by the `reply` event and matched to a command by id.
export interface ReplyData {
  id: string;
  ok: boolean;
  error?: BridgeError;
}

// Every native → web message is a uniform `{ event, data }` envelope.
export type Inbound =
  | { event: "reply"; data: ReplyData }
  | { event: "state"; data: DeviceState }
  | { event: "reading"; data: ReadingEvent }
  | { event: "capture"; data: Capture }
  | { event: "frame"; data: WireFrameEvent }
  | { event: "error"; data: BridgeError };

/** Parse a numeric value the SDK emits as a string, tolerating transient junk. */
export function parseNumeric(raw: string | null | undefined): number | null {
  if (raw == null) return null;
  const n = Number(raw.trim());
  return Number.isFinite(n) ? n : null;
}

/** Parse ECG `ecgData` into its dimensionless sample values. */
export function parseEcgData(raw: string | null | undefined): number[] {
  if (!raw) return [];
  return raw
    .trim()
    .split(/\s+/)
    .map(Number)
    .filter((n) => Number.isFinite(n));
}
